src/c_api/c_api.cc

```diff
@@ -581,19 +582,39 @@ int MXLoadLib(const char *path) {
 
     // FGradient register lambda
     auto grad_reg = [=](const nnvm::ObjectPtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
-        // copy gradients first
-        std::vector<nnvm::NodeEntry> heads(ograds.begin(), ograds.end());
-        // copy inputs second
-        for (auto& h : n->inputs) {
-          heads.push_back(h);
-        }
-        // copy outputs last
-        uint32_t n_out = n->num_outputs();
-        for (uint32_t i = 0; i < n_out; ++i) {
-          heads.emplace_back(n, i, 0);
-        }
-        std::string grad_name = "_backward_" + name_str;
-        return mxnet::op::MakeGradNode(grad_name.c_str(), n, heads, n->attrs.dict);
+      // create node for gradient
+      auto p = nnvm::Node::Create();
+      std::string grad_name = "_backward_" + name_str;
+      p->attrs.op = nnvm::Op::Get(grad_name.c_str());
+      p->attrs.name = n->attrs.name + "_backward";
```

Comment:
why do we have two different names - `grad_name` and `p->attrs.name` ?

Reply:
`grad_name` is the name of the registered FGradient operator which is: `"_backward_" + name_str`. it will be the same for all backward nodes of that particular op (think of it as backward op_name)
`p->attrs.name` is the unique name of the particular node in the graph, which is: `n->attrs.name + "_backward"`. It will be different for backward nodes of the same op

Reply:
Okay, why can't `n->attrs.name` be same for two different nodes of same operator?

Reply:
node names have to be unique

Reply:
Okay, got it. Thanks for the clarification.